src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,168 @@
+        } else if (isValidIndex(index)) {
+            Entry<T> insertEntry = new Entry<T>(value, null, null);
 
+            Entry<T> currentWithIndex = getEntryByIndex(index);
+            if (currentWithIndex != null) {
+                currentWithIndex.prev = insertEntry;
+            }
+
+            Entry<T> previousEntry = getEntryByIndex(index - 1);
+            if (previousEntry != null) {
+                previousEntry.next = insertEntry;
+            }
+
+            insertEntry.next = currentWithIndex;
+            insertEntry.prev = previousEntry;
+
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
+        if (isValidIndex(index)) {
+            Entry<T> entry = getEntryByIndex(index);
+            if (entry != null) {
+                return entry.value;
+            }
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        if (isValidIndex(index)) {
+            Entry<T> entry = getEntryByIndex(index);
+            if (entry != null) {
+                T oldvalue = entry.value;
+                entry.value = value;
+                return oldvalue;
+            }
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        if (isValidIndex(index)) {
+            Entry<T> entry = getEntryByIndex(index);
+            if (entry != null) {
+                Entry<T> entryPrevious = entry.prev;
+                if (entryPrevious != null) {
+                    entryPrevious.next = entry.next;
+                } else {
+                    head = entry.next;
+                }
+
+                Entry<T> entryNext = entry.next;
+                if (entryNext != null) {
+                    entryNext.prev = entry.prev;
+                }
+
+                size--;
+                return entry.value;
+            }
+        }
         return null;
     }
 
     @Override
     public boolean remove(T t) {
+        Entry<T> entry = head;
+        for (int i = 0; i < size; i++) {
+            if (t == entry.value || (entry.value != null && entry.value.equals(t))) {
+                remove(i);
+                return true;
+            }
+            entry = entry.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isValidIndex(int index) {
+        if (index < 0 || (index >= size)) {
+            throw new IndexOutOfBoundsException("Index [" + index + "] is out of size: " + size);
+        }
+        return true;
+    }
+
+    private Entry<T> getEntryByIndex(int index) {
+        if (isValidIndex(index)) {
```

Comment:
in this method let's start moving from the first Entry OR from the last one
depending on whether index is closer to the start or to the end’

Reply:
Done